xiaomi_ir (python-miio ChuangmiIr device)

cnrd opened the pull request:
## Description:
I wasn't really sure what to name it, it uses the xiaomi_miio package, but xiaomi_miio is already taken in switches.

Everything is tested and working, not sure if someone can come up with a better name before I start on the docs?

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#<home-assistant.github.io PR number goes here>

## Example entry for `configuration.yaml` (if applicable):
```yaml
switch:
  - platform: xiaomi_ir
    host: 192.0.1.42
    token: 1234xyz
    switches: <--- Optional
      some_switch:
        friendly_name: "Some Switch"
        command_on: "base64 package to send"
        command_off: "base64 package to send"
```

## Checklist:
  - [x] The code change is tested and works locally.

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

cnrd committed:
first working version

cnrd committed:
woof!

cnrd committed:
woof?

cnrd committed:
added requirements

cnrd committed:
added requirements, again?

cnrd committed:
Changed functions to be async, removed unused variables and config

cnrd committed:
Added file to .coveragerc

cnrd committed:
schedule update async

syssi commented on homeassistant/components/switch/xiaomi_ir.py:
```diff
@@ -0,0 +1,181 @@
+    def _send_packet(call):
+        """Send a packet."""
+        packets = call.data.get('packet', [])
+        for packet in packets:
```
If the list is not needed the for loop can be removed.

syssi commented on homeassistant/components/switch/xiaomi_ir.py:
```diff
@@ -0,0 +1,181 @@
+                    remote.play, payload, None)
+            except DeviceException as ex:
+                _LOGGER.error(
+                    "Failed to send packet, %s, exception: %s", payload, ex)
```
Transmit infrared command failed.

syssi commented on homeassistant/components/switch/xiaomi_ir.py:
```diff
@@ -0,0 +1,181 @@
+        _LOGGER.error("Token is wrong : %s", ex)
+        return
+
+    hass.services.register(DOMAIN, SERVICE_LEARN + '_' +
```
Please apply a schema for input validation.

cnrd replied:
> Please apply a schema for input validation.
Input is validated on line 37, this just makes the error message easier to read for people in case they use the wrong token (but token has the correct length).

syssi commented on homeassistant/components/switch/xiaomi_ir.py:
```diff
@@ -0,0 +1,181 @@
+    hass.services.register(DOMAIN, SERVICE_LEARN + '_' +
+                           host.replace('.', '_'), _learn_command)
+    hass.services.register(DOMAIN, SERVICE_SEND + '_' +
+                           host.replace('.', '_'), _send_packet)
```
Don't append the host to the name of the service. Address the proper entity by the entity_id.

cnrd replied:
An entity is not necessarily available, this is also a legal config:
```yaml
switch:
  - platform: xiaomi_ir
    host: some_ip
    token: 'atoken'
```

syssi commented on homeassistant/components/switch/xiaomi_ir.py:
```diff
@@ -0,0 +1,181 @@
+        from miio import DeviceException
+
+        if packet is None:
+            _LOGGER.debug("Empty packet")
```
Empty infrared command skipped.

syssi commented on homeassistant/components/switch/xiaomi_ir.py:
```diff
@@ -0,0 +1,181 @@
+            self.async_schedule_update_ha_state()
+
+    @asyncio.coroutine
+    def _sendpacket(self, packet):
```
_send_command(self, command):

syssi commented:
The prefix "xiaomi_miio" should be used here.

syssi commented on homeassistant/components/switch/xiaomi_ir.py:
```diff
@@ -0,0 +1,181 @@
+    def _learn_command(call):
+        """Handle a learn command."""
+        from random import randint
+        slot = randint(1, 1000000)
```
Please don't use a random number here. The learned command will be stored the at device at a unknown slot and can destroy used slots. Please default to a single slot and provide a service parameter which allows the selection of a specific slot between 1 and 1000000.

cnrd commented on homeassistant/components/switch/xiaomi_ir.py:
```diff
@@ -0,0 +1,181 @@
+    @asyncio.coroutine
+    def _send_packet(call):
+        """Send a packet."""
+        packets = call.data.get('packet', [])
```
There are cases where multiple button presses are needed, we have two options:
```yaml
script:
  some_script:
    sequence:
      - service: switch.chuangmiir_send_packet_some_ip
        data:
          packet:
            - 'base64'
            - 'base64'
```
or
```yaml
script:
  some_script:
    sequence:
      - service: switch.chuangmiir_send_command_some_ip
        data:
          command:
            'base64'
      - service: switch.chuangmiir_send_command_some_ip
        data:
          command:
            'base64
```

syssi replied:
I agree. A list of commands will be fine and clear if a service description is available.

cnrd replied:
I do however agree that command is a better naming scheme:
```yaml
script:
  some_script:
    sequence:
      - service: switch.chuangmiir_send_packet_some_ip
        data:
          command:
            - 'base64'
            - 'base64'
```

cnrd commented:
If we want to use switch.xiaomi_miio_infrared should the code be merged with switches/xiaomi-miio.py or should it reside in a new file: xiaomi-miio-infrared.py?

cnrd commented on homeassistant/components/switch/xiaomi_ir.py:
```diff
@@ -0,0 +1,181 @@
+
+        _LOGGER.info("Press the key you want Home Assistant to learn")
+        start_time = utcnow()
+        while (utcnow() - start_time) < timedelta(seconds=20):
```
How short would you want it to be? People do need time to press the button, run to the IR receiver and press a remote button.

cnrd committed:
Name changes, configurable slot and timeout, default timeout decreased to 10 seconds, added service descriptions, changed logger messages

cnrd commented:
@syssi I implemented most of the changes that you requested, I am however unsure of how to deal with the service calls, as it is valid to have a device configured with no configured switches, this means that no entity_ids are available.

Also I haven't touched the naming of files yet, still waiting on @balloob @rytilahti an what to do here.

thundergreen commented:
What about building a climate component in The same way? Would love it

thundergreen commented:
Maybe with or learning and standard commands .I'm waiting this for month. I bought very expensive zwave climate extenders but I think it should be possible aki to make the same like the broad link climate component

thundergreen commented:
Maybe with or learning and standard commands .I'm waiting this for month. I bought very expensive zwave climate extenders but I think it should be possible aki to make the same like the broad link climate component

cnrd commented:
@thundergreen please open a feature request, PRs should not be used for feature requests.

syssi commented on homeassistant/components/switch/xiaomi_ir.py:
```diff
@@ -0,0 +1,190 @@
+    @asyncio.coroutine
+    def _learn_command(call):
+        """Handle a learn command."""
+        slot = config.get(CONF_SLOT)
```
The service could be extended by the "slot" parameter, too: slot = call.data.get('slot', config.get(CONF_SLOT))

syssi commented on homeassistant/components/switch/services.yaml:
```diff
@@ -30,3 +30,13 @@ mysensors_send_ir_code:
+
+xiaomi_miio_learn_command:
+  description: Learn an IR command, press "Call Service", point the remote at the IR device, and the learned command will be shown as a notification in Overview.
+
```
Please add the suggested slot parameter.

syssi commented on homeassistant/components/switch/xiaomi_ir.py:
```diff
@@ -0,0 +1,190 @@
+            packet = yield from hass.async_add_job(
+                remote.read, slot)
+            if packet['code']:
+                log_msg = "Recieved command is: {}".\
```
Typo: Recieved -> Received

cnrd committed:
Fixed typo, added slot config to service, updated services.yaml

syssi commented:
It's possible to provide a schema per service. Compare:

https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/fan/xiaomi_miio.py#L142
https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/fan/xiaomi_miio.py#L142

cnrd committed:
Forgot to run linter, woof!

cnrd commented:
@syssi Would that be as a replacement for switch.xiaomi_miio_learn_command_ip_address and switch.xiaomi_miio_send_command_ip_address, so that there is are only two services: switch.xiaomi_miio_learn_command and switch.xiaomi_miio_send_command ?

syssi commented:
No. It's just a hint how to validate the parameters of a service easily.

cnrd commented:
@syssi Okay so there are a few things that we still need to address:
1. Naming of the services: What can I do here when there are no entities? (It is possible to have multiple IR devices, how do we distinguish between them)
2. Naming of the actual platform.

I'm working on that schema thing now.

cnrd committed:
Use scheme to validate service input, make setup_platform async

rytilahti commented on homeassistant/components/switch/xiaomi_ir.py:
```diff
@@ -0,0 +1,181 @@
+Support for the Xiaomi IR Remote (Chuangmi IR).
+
+For more details about this platform, please refer to the documentation
+https://home-assistant.io/components/ir_remote.xiaomi_miio/
```
Shouldn't this actually be `remote.xiaomi_miio` and use `data` instead of `command` for payloads accordingly?

rytilahti commented on homeassistant/components/switch/xiaomi_ir.py:
```diff
@@ -0,0 +1,196 @@
+SWITCH_SCHEMA = vol.Schema({
+    vol.Optional(CONF_COMMAND_OFF, default=None): cv.string,
+    vol.Optional(CONF_COMMAND_ON, default=None): cv.string,
+    vol.Optional(CONF_FRIENDLY_NAME): cv.string,
```
Are `on` and `off` the only possible commands? Or should this be done similarly to https://home-assistant.io/components/remote.itach/ with named commands?

cnrd replied:
This feels much better! On and Off really does feel weird when you can send any arbitrary IR command

rytilahti replied:
I still think that this should follow the same API as the rest of remotes, there's always `send_command` for any custom commands, if I'm understanding it correctly now.

rytilahti commented on homeassistant/components/switch/xiaomi_ir.py:
```diff
@@ -0,0 +1,196 @@
+def async_setup_platform(hass, config, async_add_devices, discovery_info=None):
+    """Set up the Xiaomi IR Remote (Chuangmi IR) platform."""
+    from miio import ChuangmiIr, DeviceException
+    from construct import ChecksumError
```
When can `ChecksumError` happen and should not this be something to handle inside python-miio?

cnrd replied:
It happens if the token is the correct length but wrong.

rytilahti replied:
I think python-miio should be patched to emit `DeviceException` or (`IncorrectTokenException` which inherits from `DeviceException`), which should then be catched here.

rytilahti commented on homeassistant/components/switch/xiaomi_ir.py:
```diff
@@ -0,0 +1,196 @@
+        packets = call.data.get('command', [])
+        for packet in packets:
+            payload = str(packet)
+            _LOGGER.debug(payload)
```
Please use descriptive log messages, e.g. `Sending payload '%s' % payload`

cnrd replied:
Done

rytilahti commented on homeassistant/components/switch/xiaomi_ir.py:
```diff
@@ -0,0 +1,196 @@
+
+
+@asyncio.coroutine
+def async_setup_platform(hass, config, async_add_devices, discovery_info=None):
```
I'm not sure if netdisco supports autodetecting these at the moment, but it may be worth to add support there and handle the case when `discovery_info` is available.

cnrd replied:
I just ran `mirobo discover` this is what came back:
```bash
INFO:miio.discovery:Discovering devices with mDNS, press any key to quit...
INFO:miio.discovery:Found a supported 'ChuangmiIr' at CORRECT_IP - token: VALID_TOKEN_SAME_AS_EXTRACTED```

rytilahti replied:
Ok, good! I'm unsure how to deal with the token, but here's an example of how `_miio._udp` yeelights are discovered over mdns in netdisco (which is used by homeassistant for discovery): https://github.com/home-assistant/netdisco/blob/master/netdisco/discoverables/yeelight.py . Maybe there should be a generic `miio.py` discoverable (there are also plenty of other devices currently not autodetected, partially because of the token issue).

rytilahti commented on homeassistant/components/switch/xiaomi_ir.py:
```diff
@@ -0,0 +1,196 @@
+    async_add_devices(switches)
+
+
+class ChuangmiIrSwitch(SwitchDevice):
```
If this fits better to `remote`, then the base class has to be changed.

rytilahti commented on homeassistant/components/switch/xiaomi_ir.py:
```diff
@@ -0,0 +1,196 @@
+            self.async_schedule_update_ha_state()
+
+    @asyncio.coroutine
+    def _sendcommand(self, command):
```
`_send_command` woudl feel more pythonic.

cnrd replied:
Done

rytilahti commented on homeassistant/const.py:
```diff
@@ -149,6 +149,7 @@
 CONF_WHITE_VALUE = 'white_value'
 CONF_XY = 'xy'
 CONF_ZONE = 'zone'
+CONF_SLOT = 'slot'
```
I'm unsure how well this would be accepted, are there other examples for its use, or could this be either defined in the platform or renamed to something else that's already in use?

cnrd replied:
I can't find anything that describes the same thing in const.py, I just had the understanding that they had to be present there to be usable in configurations. I can change it to be local to the platform instead.

rytilahti replied:
They are present there to make things consistent between different platforms exposing the same settings, I suppose.

rytilahti reviewed (commented):
> Naming of the services: What can I do here when there are no entities? (It is possible to have multiple IR devices, how do we distinguish between them)

Isn't there one instance per remote in these cases, so `entity_id` is usual should work just fine?

@syssi regarding to the naming, see my comment on `remote` :-)

cnrd commented:
@rytilahti 
> Isn't there one instance per remote in these cases, so entity_id is usual should work just fine?

Maybe I just have problems understanding how, but if the config is defined as follows:

```yaml
switch:
  - platform: xiaomi_ir
    host: [ip]
    token: [token]
```

No entities show up in hass, so I guess that I would need a way to add an entity that would only show up in dev-states, or another way to link the service call and actual object.

Not sure if something akin to this would work: https://github.com/home-assistant/home-assistant/blob/9a3895c79e2f787c71e190adda2ce1b8b0e5d5fd/homeassistant/components/vacuum/xiaomi_miio.py#L103
And then in the service call we could somehow access that object.

rytilahti commented:
> No entities show up in hass (as there are not actual switches), so I guess that I would need a way to add an entity that would only show up in dev-states, or another way to link the service call and actual object.

I suppose having it as a remote would fix this issue, but I have no idea really so someone else should fill in!

cnrd committed:
made it more pythonic

cnrd committed:
Reverted const.py changes

cnrd commented:
Okay, so I reverted the const.py change, and changed some names.

I feel like this would be a better fit for the remote type, so i'm going to attempt to refactor it into a remote device.

cnrd commented:
Too many changes, so I created a new PR: https://github.com/home-assistant/home-assistant/pull/11856
